toopher/__init__.py

```diff
@@ -238,55 +236,244 @@ def _parse_request_error(self, content):
+        else:
+            return self._raw_data[name]
+
+    def refresh_from_server(self, api):
+        url = '/user_terminals/' + self.id
+        result = api.advanced.raw.get(url)
+        self.update(result)
+
+    def update(self, json_response):
+        try:
+            self.id = json_response['id']
+            self.name = json_response['name']
+            self.name_extra = json_response['name_extra']
+            self.user = User(json_response['user'])
+        except Exception:
+            raise ToopherApiError("Could not parse user terminal from response")
+
+        self._raw_data = json_response
+
+
+class Users(object):
+    def __init__(self, raw):
+        self.raw = raw
+
+    def create(self, username, **kwargs):
+        url = '/users/create'
+        params = {'name': username}
+        params.update(kwargs)
+        result = self.raw.post(url, **params)
+        return User(result)
+
+    def get_by_id(self, user_id):
+        url = '/users/' + user_id
+        result = self.raw.get(url)
+        return User(result)
+
+    def get_by_name(self, username):
+        url = '/users'
+        users = self.raw.get(url, user_name=username)
+
+        if len(users) > 1:
+            raise ToopherApiError('Multiple users with name = %s' % username)
+        elif not len(users):
+            raise ToopherApiError('No users with name = %s' % username)
+
+        return User(users[0])
+
+
+class User(object):
+    def __init__(self, json_response):
+        self.update(json_response)
 
     def __getattr__(self, name):
         if name.startswith('__') or name not in self._raw_data:  # Exclude 'magic' methods to allow for (un)pickling
-            return super(AuthenticationStatus, self).__getattr__(name)
+            return super(User, self).__getattr__(name)
         else:
             return self._raw_data[name]
 
+    def refresh_from_server(self, api):
+        url = '/users/' + self.id
+        result = api.advanced.raw.get(url)
+        self.update(result)
+
+    def enable(self, api):
+        url = '/users/' + self.id
+        result = api.advanced.raw.post(url, disable_toopher_auth=False)
+        self.update(result)
+
+    def disable(self, api):
+        url = '/users/' + self.id
+        result = api.advanced.raw.post(url, disable_toopher_auth=True)
+        self.update(result)
+
+    def reset(self, api):
+        url = '/users/reset'
+        params = {'name': self.name}
+        api.advanced.raw.post(url, **params)
+        return True # would raise error in _request if failed
+
+    def update(self, json_response):
+        try:
+            self.id = json_response['id']
+            self.name = json_response['name']
+            self.disable_toopher_auth = json_response['disable_toopher_auth']
+        except Exception:
+            raise ToopherApiError("Could not parse user from response")
+
+        self._raw_data = json_response
+
 
 class ToopherApiError(Exception): pass
```

Comment:
Why is this repeated down here? Copy/paste mistake from awhile ago? I'm going to go ahead and delete it.